pkgs/os-specific/linux/kernel/structured-config.nix

```diff
@@ -0,0 +1,476 @@
+# Configuration
+{ lib, version, system, is64bit }:
+
+# Optional features
+{ grsecurity ? false, xen_dom0 ? false }:
+
+with lib; let 
+  # Borrowed from aszlig here: https://github.com/openlab-aux/vuizvui/blob/master/modules/user/aszlig/system/kernel.nix#L8-L18
+  generateKConf = exprs: let
+    isNumber = c: elem c ["0" "1" "2" "3" "4" "5" "6" "7" "8" "9"];
+    mkValue = val:
+      if val == "" then "\"\""
+      else if val == "y" || val == "m" || val == "n" then val
+      else if all isNumber (stringToCharacters val) then val
+      else if substring 0 2 val == "0x" then val
+      else "\"${val}\"";
+    mkConfigLine = key: val: if val == null then "" else "${key}=${mkValue val}\n";
+    mkConf = cfg: concatStrings (mapAttrsToList mkConfigLine cfg);
+  in mkConf exprs;
+
+  flattenKConf = nested: mapAttrs (_: head) (zipAttrs (attrValues nested));
+
+  # Common patterns
+  when        = cond: opt: if cond then opt else null;
+  whenAtLeast = ver: when (versionAtLeast version ver);
+  whenOlder   = ver: when (versionOlder version ver);
+  whenBetween = verLow: verHigh: when (versionAtLeast version verLow && versionOlder version verHigh);
+
+  options = {
+    debug = {
+      "DEBUG_KERNEL"               = "y";
+      "TIMER_STATS"                = "y";
+      "BACKTRACE_SELF_TEST"        = "n";
+      "CPU_NOTIFIER_ERROR_INJECT?" = "n";
+      "DEBUG_DEVRES"               = "n";
+      "DEBUG_NX_TEST"              = "n";
+      "DEBUG_STACK_USAGE"          = "n";
+      "DEBUG_STACKOVERFLOW"        = when (!grsecurity) "n";
+      "RCU_TORTURE_TEST"           = "n";
+      "SCHEDSTATS"                 = "n";
+      "DETECT_HUNG_TASK"           = "y";
+      "CRASH_DUMP?"                = "n";
```

Comment:
Putting `?` in the option names is probably a bad idea because it would complicate merging values. E.g. one module might set `CRASH_DUMP?` and another `CRASH_DUMP`. It might be better to move the `?` to the right-hand side, e.g.
```
CRASH_DUMP = "?n";
```
or in a structured way
```
CRASH_DUMP = ifSupported "n";
```